packages/router-core/src/router.ts

```diff
@@ -2179,588 +2179,567 @@ export class RouterCore<
-                  if (parentMatch?.ssr === false) {
-                    ssr = false
-                  } else {
-                    let tempSsr: boolean | 'data-only'
-                    if (route.options.ssr === undefined) {
-                      tempSsr = defaultSsr
-                    } else if (typeof route.options.ssr === 'function') {
-                      const { search, params } = this.getMatch(matchId)!
-
-                      function makeMaybe(value: any, error: any) {
-                        if (error) {
-                          return { status: 'error' as const, error }
-                        }
-                        return { status: 'success' as const, value }
-                      }
-
-                      const ssrFnContext: SsrContextOptions<any, any, any> = {
-                        search: makeMaybe(search, existingMatch.searchError),
-                        params: makeMaybe(params, existingMatch.paramsError),
-                        location,
-                        matches: matches.map((match) => ({
-                          index: match.index,
-                          pathname: match.pathname,
-                          fullPath: match.fullPath,
-                          staticData: match.staticData,
-                          id: match.id,
-                          routeId: match.routeId,
-                          search: makeMaybe(match.search, match.searchError),
-                          params: makeMaybe(match.params, match.paramsError),
-                          ssr: match.ssr,
-                        })),
-                      }
-                      tempSsr =
-                        (await route.options.ssr(ssrFnContext)) ?? defaultSsr
-                    } else {
-                      tempSsr = route.options.ssr
-                    }
+        const pendingMs =
+          route.options.pendingMs ?? this.options.defaultPendingMs
 
-                    if (tempSsr === true && parentMatch?.ssr === 'data-only') {
-                      ssr = 'data-only'
-                    } else {
-                      ssr = tempSsr
-                    }
+        // on the server, determine whether SSR the current match or not
+        if (this.isServer) {
+          let ssr: boolean | 'data-only'
+          // in SPA mode, only SSR the root route
+          if (this.isShell()) {
+            ssr = matchId === rootRouteId
+          } else {
+            const defaultSsr = this.options.defaultSsr ?? true
+            if (parentMatch?.ssr === false) {
+              ssr = false
+            } else {
+              let tempSsr: boolean | 'data-only'
+              if (route.options.ssr === undefined) {
+                tempSsr = defaultSsr
+              } else if (typeof route.options.ssr === 'function') {
+                const { search, params } = this.getMatch(matchId)!
+
+                function makeMaybe(value: any, error: any) {
```

Comment:
Would the definition of `makeMaybe` outside of the for loop have any performance gains? (still being left within the scope of `loadMatches` though)